pkg/bisect/generic/slice.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package generic
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type MinimizeConfig[T any] struct {
+	// The original slice is minimized with respect to this predicate.
+	// If Pred(X) returns true, X is assumed to contain all elements that must stay.
+	Pred func([]T) (bool, error)
+	// MaxSteps is a limit on the number of predicate calls during bisection.
+	// If it's hit, the bisection continues as if Pred() begins to return false.
+	// If it's set to 0 (by default), no limit is applied.
+	MaxSteps int
+	// MaxChunks sets a limit on the number of chunks pursued by the bisection algorithm.
+	// If we hit the limit, bisection is stopped and Array() returns ErrTooManyChunks
+	// anongside the intermediate bisection result (a valid, but not fully minimized slice).
+	MaxChunks int
+	// Logf is used for sharing debugging output.
+	Logf func(string, ...interface{})
+}
+
+// MinimizeSlice finds a minimal subsequence of slice elements that still gives Pred() == true.
+// The algorithm works by sequentially splitting the slice into smaller-size chunks and running
+// Pred() witout those chunks. MinimizeSlice() receives the original slice chunks.
+// The expected number of Pred() runs is O(|result|*log2(|elements|)).
+func MinimizeSlice[T any](config MinimizeConfig[T], startChunks ...[]T) ([]T, error) {
+	if config.Logf == nil {
+		config.Logf = func(string, ...interface{}) {}
+	}
+	var chunks []*arrayChunk[T]
+	for _, elems := range startChunks {
+		if len(elems) == 0 {
+			continue
+		}
+		chunks = append(chunks, &arrayChunk[T]{
+			elements: elems,
+		})
+	}
+	ctx := &sliceCtx[T]{
+		config: config,
+		chunks: chunks,
+	}
+	return ctx.bisect()
+}
+
+type sliceCtx[T any] struct {
+	config   MinimizeConfig[T]
+	chunks   []*arrayChunk[T]
+	predRuns int
+}
+
+type arrayChunk[T any] struct {
+	elements []T
+	final    bool // There's no way to further split this chunk.
+}
+
+// ErrTooManyChunks is returned if the number of necessary chunks surpassed MaxChunks.
+var ErrTooManyChunks = errors.New("the bisection process is following too many necessary chunks")
+
+func (ctx *sliceCtx[T]) bisect() ([]T, error) {
+	// At first, we don't know if the original chunks are really necessary.
+	err := ctx.dropIndividualChunks()
```

Comment:
We still need to make sure that all original chunks are necessary. It reduces the number of predicate runs needed later, see https://github.com/google/syzkaller/pull/3970#discussion_r1234282695

Though if there's just one chunk, maybe it makes more sense to run the predicate with the full input first (instead of running it with an empty array).